File: app/src/main/java/com/example/architechturestartercode/data/movie/datasource/remote/MoviesRemoteDataSource.java
package com.example.architechturestartercode.data.movie.datasource.remote;

import com.example.architechturestartercode.data.movie.model.MovieResponse;
import com.example.architechturestartercode.data.network.Network;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MoviesRemoteDataSource {
    private MoviesService moviesService;

    public MoviesRemoteDataSource() {
        this.moviesService = Network.getInstance().getMoviesService();
    }

    public void getAllMovies(MoviesNetworkResponse callback) {
        moviesService.getMovies().enqueue(new Callback<MovieResponse>() {
            @Override
            public void onResponse(Call<MovieResponse> call, Response<MovieResponse> response) {
                if (response.code() == 200) {
                    callback.onSuccess(response.body().results);
                } else {
                    callback.serverError("Error code: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<MovieResponse> call, Throwable t) {
                if (t instanceof java.io.IOException) {
                    callback.onFailure("Network failure, please try again");
                } else {
                    callback.onFailure("Conversion issue! Big problems :(");
                }
            }
        });
    }

}

File: app/src/main/java/com/example/architechturestartercode/data/movie/datasource/remote/MoviesRemoteDataSource.kt
package com.example.architechturestartercode.data.movie.datasource.remote

import com.example.architechturestartercode.data.movie.datasource.remote.model.RemoteMovie
import javax.inject.Inject

class MoviesRemoteDataSource @Inject constructor(private val moviesService: MoviesService) :
    IMoviesRemoteDataSource {


    override suspend fun getAllMovies(): Result<List<RemoteMovie>> {
        val response = moviesService.getMovies()
        if (response.isSuccessful) {
            val movies = response.body()?.results ?: emptyList()
            return Result.success(movies)
        } else {
            val e = Exception("Failed to fetch movies: ${response.code()} ${response.message()}")
            return Result.failure(e)
        }
    }
}

